File: ebics-web-ui/src/components/user-settings.ts
import { ref, onMounted } from 'vue';
import { UserSettings } from 'components/models';
import { api } from 'boot/axios';
import useBaseAPI from './base-api';

const userSettings = ref<UserSettings>({
  uploadOnDrop: true,
  testerSettings: true,
  adjustmentOptions: {
    applyAutomatically: true,
    pain00x: {
      msgId: true,
      pmtInfId: false,
      instrId: true,
      endToEndId: true,
      uetr: true,
      nbOfTrxsCalc: true,
      ctrlSumCalc: true,
      reqdExctnDt: true,
      creDtTm: true,
      idPrefix: 't51246',
    },
    swift: {
      uetr: true,
      f20: true,
      f21: false,
      f30: true,
      idPrefix: 't111111',
      randomIds: false,
    },
  },
  displayAdminTypes: false,
});

/**
 * 
 * @returns user settings related API
 */
export default function useUserSettingsAPI() {
  const { apiErrorHandler, apiOkHandler } = useBaseAPI();

  const loadUserSettings = async (): Promise<void> => {
    try {
      const response = await api.get<UserSettings>('user/settings');
      userSettings.value = response.data;
    } catch (error) {
      apiErrorHandler('Loading of user settings failed', error);
    }
  };

  const saveUserSettings = async (): Promise<void> => {
    try {
      await api.put<UserSettings>('user/settings', userSettings.value);
      apiOkHandler('User settings successfully saved')
    } catch (error) {
      apiErrorHandler('Saving settings failed', error);
    }
  };

  onMounted(loadUserSettings);

  return { userSettings, loadUserSettings, saveUserSettings };
}

File: ebics-dbmodel/src/main/kotlin/org/ebics/client/api/bankconnection/settings/UserSettings.kt
package org.ebics.client.api.bankconnection.settings

import javax.persistence.*

data class UserSettingsData(
    val uploadOnDrop: Boolean,
    val testerSettings: Boolean,
    val adjustmentOptions: AdjustmentOptions,
)

/**
    User Settings class
 */
@Entity
class UserSettings(
    @Id
    val userId: String,

    val uploadOnDrop: Boolean,
    val testerSettings: Boolean,

    @Embedded
    val adjustmentOptions: AdjustmentOptions,
) : UserSettingsAccessRightsController {
    override fun getOwnerName(): String = userId
    override fun getObjectName(): String = "User settings of '$userId'"
}

@Embeddable
class AdjustmentOptions(
    val applyAutomatically: Boolean,

    @Embedded
    val pain00x: AdjustmentsOptionsPain00x,

    @Embedded
    val swift: AdjustmentsOptionsSwift,
)

@Embeddable
class AdjustmentsOptionsPain00x(
    val msgId: Boolean,
    val pmtInfId: Boolean,
    val instrId: Boolean,
    val endToEndId: Boolean,
    val uetr: Boolean,
    val reqdExctnDt: Boolean,
    val creDtTm: Boolean,
    val nbOfTrxsCalc: Boolean,
    val ctrlSumCalc: Boolean,
    val idPrefix: String,
)

@Embeddable
class AdjustmentsOptionsSwift(
    val uetr: Boolean,
    val f20: Boolean,
    val f21: Boolean,
    val randomIds: Boolean,
    val f30: Boolean,
    val idPrefix: String,
)